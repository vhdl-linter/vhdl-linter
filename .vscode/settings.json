{
  "typescript.tsdk": "node_modules/typescript/lib",
  "eslint.lintTask.enable": true,
  "eslint.lintTask.options": "--ext .ts .",
  "eslint.format.enable": true,
  "cSpell.words": [
    "actuals",
    "clientrc",
    "downto",
    "Inoutput",
    "Mentionables",
    "nand",
    "Onig",
    "OSVVM",
    "quickfix",
    "Referencable",
    "setq",
    "subarchitecture",
    "Textmate",
    "ulogic",
    "uninstantiated",
    "Uniquify",
    "Verilog",
    "vhdl",
    "Vhdl",
    "vmode",
    "vprop",
    "vunit",
    "Whitespaces",
    "xnor"
  ]
}