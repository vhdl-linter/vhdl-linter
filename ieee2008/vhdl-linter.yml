rules:
  type-resolved: false
  unit: false