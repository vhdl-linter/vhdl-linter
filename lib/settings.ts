import { DeepPartial } from 'utility-types';
export interface ISettings {
  ports: {
    outRegex: string;
    inRegex: string;
    enablePortStyle: boolean;
  };
  paths: {
    additional: string[];
    ignoreRegex: string;
  };
  style: {
    preferredLogicTypePort: "unresolved" |  "resolved" | "ignore";
    preferredLogicTypeSignal: "unresolved" | "resolved" | "ignore";
    unusedSignalRegex: string;
    ieeeCasing: 'lowercase' | 'UPPERCASE';
  };
  rules: {
    warnLibrary: boolean;
    warnLogicType: boolean;
    warnMultipleDriver: boolean;
  };
  semanticTokens: boolean;
}
export const defaultSettings: ISettings = {
  ports: {
    outRegex: '^o_',
    inRegex: '^i_',
    enablePortStyle: true,
  },
  paths: {
    additional: [],
    ignoreRegex: ''
  },
  style: {
    preferredLogicTypePort: 'unresolved',
    preferredLogicTypeSignal: 'unresolved',
    unusedSignalRegex: '_unused$',
    ieeeCasing: 'lowercase'
  },
  rules: {
    warnLogicType: true,
    warnLibrary: false,
    warnMultipleDriver: false
  },
  semanticTokens: false
};
export function defaultSettingsGetter() {
  return defaultSettings;
}
export function defaultSettingsWithOverwrite(overwrite?: DeepPartial<ISettings>) {
  const newDefault = Object.assign({}, defaultSettings);

  if (overwrite) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    recursiveObjectAssign(newDefault as any, overwrite);
  }
  return () => defaultSettings;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function recursiveObjectAssign<T extends Record<string, any>>(target: T, source: Partial<T>) {
  Object.keys(source).forEach(key => {
    const s_val = source[key];
    const t_val = target[key];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (target as any)[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
      ? recursiveObjectAssign(t_val, s_val)
      : s_val;
  });
  return target;
}