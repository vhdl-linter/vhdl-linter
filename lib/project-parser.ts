import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises } from 'fs';
import { join, sep } from 'path';
import { cwd } from 'process';
import { OContext, OEntity, OPackage, OPackageInstantiation } from './parser/objects';
import { SettingsGetter, VhdlLinter } from './vhdl-linter';

export class ProjectParser {

  public cachedFiles: FileCache[] = [];
  public packages: OPackage[] = [];
  public packageInstantiations: OPackageInstantiation[] = [];
  public contexts: OContext[] = [];
  public entities: OEntity[] = [];
  events = new EventEmitter();
  private watchers: FSWatcher[] = [];
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(workspaces: string[], fileIgnoreRegex: string, settingsGetter: SettingsGetter, disableWatching = false) {
    const projectParser = new ProjectParser(workspaces, fileIgnoreRegex, settingsGetter);
    await projectParser.init(disableWatching);
    return projectParser;
  }
  private constructor(public workspaces: string[], public fileIgnoreRegex: string, public settingsGetter: SettingsGetter) { }
  public addFolders(pathes: string[]) {
    for (const path of pathes) {
      const watcher = watch(path.replace(sep, '/') + '/**/*.vhd?(l)', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = await (FileCache.create(path, this));
        this.cachedFiles.push(cachedFile);
        this.flattenProject();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
          this.flattenProject();
          this.events.emit('change', 'change', path);
        } else {
          console.error('modified file not found', path);
        }
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        if (cachedFileIndex > -1) {
          this.cachedFiles.splice(cachedFileIndex, 1);
          this.flattenProject();
          this.events.emit('change', 'unlink', path);
        }
      });
      this.watchers.push(watcher);
    }
  }
  private async init(disableWatching: boolean) {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    (await this.parseDirectory(join(cwd(), 'ieee2008'))).forEach(file => files.add(file));

    for (const file of files) {
      const cachedFile = await FileCache.create(file, this);
      this.cachedFiles.push(cachedFile);
    }
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.flattenProject();
    if (!disableWatching) {
      this.addFolders(this.workspaces);
    }
  }
  async stop() {
    for (const watcher of this.watchers) {
      await watcher.close();
    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private flattenProject() {
    this.packages = [];
    this.packageInstantiations = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.packageInstantiations) {
        this.packageInstantiations.push(...cachedFile.packageInstantiations);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }

}

export class FileCache {
  packages?: OPackage[];
  packageInstantiations?: OPackageInstantiation[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  text: string;
  linter: VhdlLinter;
  lintingTime: number;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(path: string, projectParser: ProjectParser) {
    const cache = new FileCache(path, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public path: string, public projectParser: ProjectParser) {
  }
  async parse() {
    this.text = await promises.readFile(this.path, { encoding: 'utf8' });
    this.linter = new VhdlLinter(this.path, this.text, this.projectParser, this.projectParser.settingsGetter);
    this.packages = this.linter.file.packages.filter((p): p is OPackage => p instanceof OPackage);
    this.packageInstantiations = this.linter.file.packageInstantiations;
    this.entities = this.linter.file.entities;
    this.contexts = this.linter.file.contexts;
  }

}