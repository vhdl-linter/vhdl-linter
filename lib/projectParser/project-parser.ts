import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises, readFileSync } from 'fs';
import { Worker } from 'node:worker_threads';
import { join, sep } from 'path';
import { OContext, OEntity, OFile, OPackage, OPackageBody } from '../parser/objects';
import * as objects from '../parser/objects';
import { VhdlLinter } from '../vhdl-linter';
// import { StaticPool } from 'node-worker-threads-pool';
import { Parser } from '../parser/parser';
import { pool } from './parser-pool';
export class ProjectParser {

  public cachedFiles: OFileCache[] = [];
  private packages: (OPackage | OPackageBody)[];
  private contexts: OContext[] = [];
  private entities: OEntity[];
  events = new EventEmitter();
  constructor(public workspaces: string[], public fileIgnoreRegex: string) { }
  async init() {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      // console.log('dir', directory);
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    // for (const directory of this.workspaces) {
    //   this.parseDirectory(directory).forEach(file => files.add(realpathSync(file)));
    // }
    const pkg = __dirname;
    if (pkg) {
      //       console.log(pkg, new Directory(pkg + '/ieee2008'));
      (await this.parseDirectory(join(pkg, `${sep}..${sep}..${sep}..${sep}ieee2008`))).forEach(file => files.add(file));
    }
    for (const file of files) {
      const cachedFile = new OFileCache(file);
      await cachedFile.parse();
      this.cachedFiles.push(cachedFile);
    }
    // this.cachedFiles.push(...await Promise.all(Array.from(files).map(async file => {
    //   const cachedFile = new OFileCache(file, this);
    //   await cachedFile.parse();
    //   return cachedFile;
    // })));
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.fetchEntitesAndPackagesAndContexts();
    for (const workspace of this.workspaces) {
      const watcher = watch(workspace.replace(sep, '/') + '/**/*.vhd', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = new OFileCache(path);
        await cachedFile.parse();
        this.cachedFiles.push(cachedFile);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
        } else {
          console.error('modified file not found', path);
        }
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'change', path);
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        this.cachedFiles.splice(cachedFileIndex, 1);
        this.fetchEntitesAndPackagesAndContexts();
        this.events.emit('change', 'unlink', path);
      });

    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private fetchEntitesAndPackagesAndContexts() {
    //     console.log(this.cachedFiles);
    this.packages = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }
  watcher: FSWatcher;
  addFolders(folders: string[]) {
    this.watcher.add(folders.map(folder => folder.replace(sep, '/') + '/**/*.vhd'));
  }
  public getPackages() {
    return this.packages;
  }
  public getContexts() {
    return this.contexts;
  }
  public getEntities() {
    return this.entities;
  }
}

export class OFileCache {

  path: string;
  digest: string;
  file: OFile;
  packages?: (OPackage | OPackageBody)[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  lintingTime: number;
  constructor(file: string) {
    // this.digest = await file.getDigest();
    this.path = file;
  }
  async parse() {
    console.log(this.path, 'start');
    // const parser = new Parser(readFileSync(this.path, { encoding: 'utf-8' }), this.path, true);
    // this.file = parser.parse();

    this.file = await pool.exec({
      path: this.path,
    });
    console.log(this.path, 'end');
    this.packages = this.file.packages;
    this.entities = this.file.entities;
    this.contexts = this.file.contexts;
  }
}