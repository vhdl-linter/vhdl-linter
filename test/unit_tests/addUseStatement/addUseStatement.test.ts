
import { expect, test } from '@jest/globals';
import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { ProjectParser } from '../../../lib/project-parser';
import { defaultSettingsGetter } from '../../../lib/settings';
import { VhdlLinter } from '../../../lib/vhdl-linter';

const files = readdirSync(__dirname).filter(file => file.endsWith('.vhd'));
test.each(files)('testing add use statement actions for file %s', async (file: string) => {
  const path = join(__dirname, file);
  const linter = new VhdlLinter(path, readFileSync(path, { encoding: 'utf8' }),
    await ProjectParser.create([__dirname], '', defaultSettingsGetter), defaultSettingsGetter);
  await linter.checkAll();

  expect(linter.messages).toMatchSnapshot();
  for (const message of linter.messages) {
    if (typeof message.code === 'string') {
      for (const action of message.code?.split(';') ?? []) {
        const actions = linter.diagnosticCodeActionRegistry[parseInt(action)]('file://dummy.vhd');
        expect(actions).toMatchSnapshot();

      }
    }
  }
});