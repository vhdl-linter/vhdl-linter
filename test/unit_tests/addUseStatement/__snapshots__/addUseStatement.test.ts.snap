// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing add use statement actions for file arch.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'std_ulogic' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 30,
        "line": 5,
      },
      "start": {
        "character": 20,
        "line": 5,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file arch.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "library ieee;
use ieee.std_logic_1164.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.std_logic_1164",
  },
]
`;

exports[`testing add use statement actions for file arch.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 5,
              },
              "start": {
                "character": 1000,
                "line": 5,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;

exports[`testing add use statement actions for file constant.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'CopyRightNotice' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 26,
        "line": 8,
      },
      "start": {
        "character": 11,
        "line": 8,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file constant.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.fixed_generic_pkg.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.fixed_generic_pkg",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.float_generic_pkg.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.float_generic_pkg",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.math_complex.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.math_complex",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.math_real.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.math_real",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_bit.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_bit",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_bit_unsigned.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_bit_unsigned",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_std.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_std",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_std_unsigned.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_std_unsigned",
  },
  {
    "command": {
      "arguments": [
        {
          "position": {
            "character": 0,
            "line": 5,
          },
          "signalName": "CopyRightNotice",
          "textDocumentUri": "file://dummy.vhd",
        },
      ],
      "command": "vhdl-linter:add-signal",
      "title": "add signal to architecture",
    },
    "kind": "quickfix",
    "title": "add signal to architecture",
  },
]
`;

exports[`testing add use statement actions for file constant.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 8,
              },
              "start": {
                "character": 1000,
                "line": 8,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;

exports[`testing add use statement actions for file entity.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'std_ulogic' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 22,
        "line": 3,
      },
      "start": {
        "character": 12,
        "line": 3,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file entity.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.std_logic_1164.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.std_logic_1164",
  },
]
`;

exports[`testing add use statement actions for file entity.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 3,
              },
              "start": {
                "character": 1000,
                "line": 3,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;

exports[`testing add use statement actions for file entityWithLibrary.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'std_ulogic' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 22,
        "line": 2,
      },
      "start": {
        "character": 12,
        "line": 2,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file entityWithLibrary.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "library ieee;
use ieee.std_logic_1164.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.std_logic_1164",
  },
]
`;

exports[`testing add use statement actions for file entityWithLibrary.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 2,
              },
              "start": {
                "character": 1000,
                "line": 2,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;

exports[`testing add use statement actions for file function.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'to_integer' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 20,
        "line": 8,
      },
      "start": {
        "character": 10,
        "line": 8,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file function.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.fixed_generic_pkg.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.fixed_generic_pkg",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.float_generic_pkg.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.float_generic_pkg",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_bit.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_bit",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_bit_unsigned.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_bit_unsigned",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_std.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_std",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "use ieee.numeric_std_unsigned.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 1,
              },
              "start": {
                "character": 0,
                "line": 1,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.numeric_std_unsigned",
  },
  {
    "command": {
      "arguments": [
        {
          "position": {
            "character": 0,
            "line": 7,
          },
          "signalName": "to_integer",
          "textDocumentUri": "file://dummy.vhd",
        },
      ],
      "command": "vhdl-linter:add-signal",
      "title": "add signal to architecture",
    },
    "kind": "quickfix",
    "title": "add signal to architecture",
  },
]
`;

exports[`testing add use statement actions for file function.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 8,
              },
              "start": {
                "character": 1000,
                "line": 8,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;

exports[`testing add use statement actions for file generic_function_call.vhd 1`] = `
[
  {
    "code": "0;1",
    "message": "signal 'ceil' is referenced but not declared (not-declared)",
    "range": {
      "end": {
        "character": 25,
        "line": 17,
      },
      "start": {
        "character": 21,
        "line": 17,
      },
    },
    "severity": 1,
  },
]
`;

exports[`testing add use statement actions for file generic_function_call.vhd 2`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "library ieee;
use ieee.math_real.all;
",
            "range": {
              "end": {
                "character": 0,
                "line": 9,
              },
              "start": {
                "character": 0,
                "line": 9,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "add use statement for ieee.math_real",
  },
  {
    "command": {
      "arguments": [
        {
          "position": {
            "character": 0,
            "line": 14,
          },
          "signalName": "ceil",
          "textDocumentUri": "file://dummy.vhd",
        },
      ],
      "command": "vhdl-linter:add-signal",
      "title": "add signal to architecture",
    },
    "kind": "quickfix",
    "title": "add signal to architecture",
  },
]
`;

exports[`testing add use statement actions for file generic_function_call.vhd 3`] = `
[
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": " -- vhdl-linter-disable-line not-declared",
            "range": {
              "end": {
                "character": 1000,
                "line": 17,
              },
              "start": {
                "character": 1000,
                "line": 17,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared on this line.",
  },
  {
    "edit": {
      "changes": {
        "file://dummy.vhd": [
          {
            "newText": "-- vhdl-linter-disable not-declared
",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
      },
    },
    "kind": "quickfix",
    "title": "Ignore not-declared for this file.",
  },
]
`;
